### ISSUE TITLE
Implement incremental update processing

### ISSUE THREAD

**Issue by azaroth42** [project-lux/data-pipeline#13]
New records are easy to process as step-wise or incremental updates, however updates to existing records and especially deletions of records are a challenge.

* Updated records might be updated such that they no longer refer to another record. If that was the last reference to it, then the previously-referenced record should be deleted as unnecessary
* For delete, if there are other internal records that have been reconciled, then simply remove the deleted record and rebuild it (e.g. it should still exist after the delete)
* However otherwise, like for updates, we need to know which other records refer to the record to determine if it should be deleted as no longer needed, or should be maintained for dataset integrity

This means tracking references between records, which has been started, but takes considerable time and network traffic to manage. Needs to be more efficient and reliable.

**Comment by azaroth42** [comment]
Incremental step towards incremental processing:

The harvest process fetches records plus the type of change. The type of change makes the main difference.
 
If the change type is:
  * Create ... then the record is new, so it won't cause downstream effects other than potentially other new records.
  * Delete, and ...
    * the LUX record is reconciled against other internal records, then the only change is to delete the equivalent (and potentially source) from the record -- e.g. if we ran everything, we'd still end up with the record from the other units.
    * and there's no other internal records, then delete from the data cache and we're eventually consistent when we rebuild from zero. To figure out how to do better in iteration 2.
  * Update, then try to determine what sort of change is in the record, and ...
    * if the change is just internal (string or number values changed, or internal structure without URIs is deleted or added) then treat it as Create.
    * if the change creates/modifies/deletes something that references another URI, then we don't know what the downstream effects will be, so store the updated record but don't process it.
 
Once harvesting is done, we should end up with a list of record ids to process. Without clearing the caches, run those records through the pipeline and track which lux records are changed, including running the references. Then all we need to do is export the changed records to a patch file to load to Marklogic.

To try and describe the delete conditions/complexity:

We want to avoid records that are not internal and have no references from internal records and are not needed to maintain consistency between external records.
 
For example...
  * Record A (to be deleted) is classified_as a concept B that's directly linked as an AAT term.
  * If A is the only reference to B, then there's no reason to have B and it should be deleted. 
  * If another Concept record C is broader to B, then B is needed to maintain the hierarchy of C and shouldn't be deleted. And the same for any other sort of reference to B.
  * If B is reconciled to another internal record, that isn't to be deleted, then it shouldn't be deleted either (as the referencing record is itself)
  * If B has no other references in, isn't an internal record itself, and refers to a record D which, after B is deleted, also has no references (etc), then D should be deleted as well, and so on.
  * If B has a reference in (and thus shouldn't be deleted) then it should still have the deleted record's URI removed from its `equivalent` property and the `source` from the administrative metadata for MarkLogic... unless there are other records from the same source in equivalent.

**Comment by azaroth42** [comment]
I think these are the different scenarios:
<img width="631" alt="scenarios" src="https://github.com/user-attachments/assets/121bfa41-acda-44a1-91e1-9d0d65986312" />

**Comment by azaroth42** [comment]
And potential solutions, which require investigation:
* Maintain the reference chains in the pipeline for records that do not have YUL URIs associated. This is more feasible as the number of records is closer to 6M than 42M. As YUL doesn't do activity streams, we won't need to calculate a delete for one of their records. Don't maintain links for core concepts like English or Primary Name even though they might not have an internal record equivalent. Possibly networkx, possibly just some big redis table. Cost is spread between initial build (construct the big table/graph) and the update processing.
* Do real time searches against LUX during the deletion to find out if there are incoming references. We would need to make sure that there's a good index for this, as not every relationship is actually searchable today. Cost is all in update processing and as it goes out over the network will be much slower, but doesn't take any additional time during initial build to construct the data.

**Comment by azaroth42** [comment]
Noting that all of the Peabody events are either Update or Delete, not Create. E.g. from today: https://images.peabody.yale.edu/data/object/masterobject0224.json

We can detect update=create based on whether there's a record in the datacache or not.

**Comment by azaroth42** [comment]
Noting for the processing, we need to do all updates and creates first, before testing for references to see if we can delete a record, as the new changes to the records might create or remove references that would allow/prevent (respectively) the delete from happening. We can then do deletes by breadth first passes, and allow ML to reindex in between each pass, to have it recalculate the triples based on the new set of records.
